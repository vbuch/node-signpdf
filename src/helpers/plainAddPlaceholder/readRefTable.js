import SignPdfError from '../../SignPdfError';
import xrefToRefMap from './xrefToRefMap';

export const getLastTrailerPosition = (pdf) => {
    const trailerStart = pdf.lastIndexOf(Buffer.from('trailer', 'utf8'));
    const trailer = pdf.slice(trailerStart, pdf.length - 6);

    const xRefPosition = trailer
        .slice(trailer.lastIndexOf(Buffer.from('startxref', 'utf8')) + 10)
        .toString();

    return parseInt(xRefPosition);
};

export const getXref = (pdf, position) => {
    let refTable = pdf.slice(position); // slice starting from where xref starts
    const realPosition = refTable.indexOf(Buffer.from('xref', 'utf8'));
    if (realPosition === -1) {
        throw new SignPdfError(
            `Could not find xref anywhere at or after ${position}.`,
            SignPdfError.TYPE_PARSE,
        );
    }
    if (realPosition > 0) {
        const prefix = refTable.slice(0, realPosition);
        if (prefix.toString().replace(/\s*/g, '') !== '') {
            throw new SignPdfError(
                `Expected xref at ${position} but found other content.`,
                SignPdfError.TYPE_PARSE,
            );
        }
    }

    const nextEofPosition = refTable.indexOf(Buffer.from('%%EOF', 'utf8'));
    if (nextEofPosition === -1) {
        throw new SignPdfError(
            'Expected EOF after xref and trailer but could not find one.',
            SignPdfError.TYPE_PARSE,
        );
    }
    refTable = refTable.slice(0, nextEofPosition);
    refTable = refTable.slice(realPosition + 4); // move ahead with the "xref"
    refTable = refTable.slice(refTable.indexOf('\n') + 1); // move after the next new line

    // extract the size
    let size = refTable.toString().split('/Size')[1];

    if (!size) {
        throw new SignPdfError(
            'Size not found in xref table.',
            SignPdfError.TYPE_PARSE,
        );
    }
    size = (/^\s*(\d+)/).exec(size);
    if (size === null) {
        throw new SignPdfError(
            'Failed to parse size of xref table.',
            SignPdfError.TYPE_PARSE,
        );
    }
    size = parseInt(size[1]);

    const [objects, infos] = refTable.toString().split('trailer');

    const isContainingPrev = infos.split('/Prev')[1] != null;

    let prev;
    if (isContainingPrev) {
        const pagesRefRegex = /Prev (\d+)/g;
        const match = pagesRefRegex.exec(infos);
        const [, prevPosition] = match;
        prev = prevPosition;
    }

    const xRefContent = xrefToRefMap(objects);

    return {
        size,
        prev,
        xRefContent,
    };
};

export const getFullXrefTable = (pdf) => {
    const lastTrailerPosition = getLastTrailerPosition(pdf);
    const lastXrefTable = getXref(pdf, lastTrailerPosition);

    if (lastXrefTable.prev === undefined) {
        return lastXrefTable.xRefContent;
    }
    const pdfWithoutLastTrailer = pdf.slice(0, lastTrailerPosition);
    const partOfXrefTable = getFullXrefTable(pdfWithoutLastTrailer);

    const mergedXrefTable = new Map([
        ...partOfXrefTable,
        ...lastXrefTable.xRefContent,
    ]);

    return mergedXrefTable;
};

/**
 * @param {Buffer} pdfBuffer
 * @returns {object}
 */
const readRefTable = (pdf) => {
    const fullXrefTable = getFullXrefTable(pdf);
    const startingIndex = 0;
    const maxIndex = Math.max(...fullXrefTable.keys());

    return {
        startingIndex,
        maxIndex,
        offsets: fullXrefTable,
    };
};

export default readRefTable;
