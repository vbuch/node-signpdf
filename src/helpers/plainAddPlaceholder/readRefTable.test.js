import fs from 'fs';
import SignPdfError from '../../SignPdfError';
import readRefTable, {getXref} from './readRefTable';

describe('getXref', () => {
    it('Throws an error when xref is not found at position', () => {
        const pdf = Buffer.from('Not containing an X R E F.');
        const position = 0;
        try {
            getXref(pdf, position);
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_PARSE);
            expect(e.message).toMatchSnapshot();
        }
    });
    it('Throws an error when xref is not at its expected position', () => {
        const pdf = Buffer.from('Contains xref but definetely not where expected.');
        const position = 2;
        try {
            getXref(pdf, position);
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_PARSE);
            expect(e.message).toMatchSnapshot();
        }
    });
    it('Throws an error when size is not found', () => {
        const pdf = Buffer.from('xref\n and there is no Size %%EOF');
        const position = 0;
        try {
            getXref(pdf, position);
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_PARSE);
            expect(e.message).toMatchSnapshot();
        }
    });
    it('Throws an error when size has unexpected value', () => {
        const pdf = Buffer.from('xref\n and then somewhere here there is a /Size XXXX %%EOF');
        const position = 0;
        try {
            getXref(pdf, position);
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_PARSE);
            expect(e.message).toMatchSnapshot();
        }
    });
    it('Throws an error when next EOF is not found', () => {
        const pdf = Buffer.from('xref\n then /Size 123 and then but no EOF');
        const position = 0;
        try {
            getXref(pdf, position);
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_PARSE);
            expect(e.message).toMatchSnapshot();
        }
    });
});

describe('readRefTable', () => {
    it('Expects to merge correctly the refTable of resources', () => {
        [
            'signed-once.pdf',
            'signed-twice.pdf',
            'contributing.pdf',
            'formexample.pdf',
            'incrementally_signed.pdf',
            'signed.pdf',
            'w3dummy.pdf',
        ].forEach((fileName) => {
            const pdf = fs.readFileSync(`${__dirname}/../../../resources/${fileName}`);
            const r = readRefTable(pdf);
            expect(r).toMatchSnapshot();
        });
    });
});
