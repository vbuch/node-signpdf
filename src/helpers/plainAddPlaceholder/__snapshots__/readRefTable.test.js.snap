// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getXref Throws an error when next EOF is not found 1`] = `"Expected EOF after xref and trailer but could not find one."`;

exports[`getXref Throws an error when size has unexpected value 1`] = `"Failed to parse size of xref table."`;

exports[`getXref Throws an error when size is not found 1`] = `"Size not found in xref table."`;

exports[`getXref Throws an error when xref is not at its expected position 1`] = `"Expected xref at 2 but found other content."`;

exports[`getXref Throws an error when xref is not found at position 1`] = `"Could not find xref anywhere at or after 0."`;

exports[`readRefTable Expects to merge correctly the refTable of resources 1`] = `
Object {
  "maxIndex": 19,
  "offsets": Map {
    1 => 19012,
    2 => 19,
    3 => 224,
    4 => 12330,
    5 => 244,
    6 => 11154,
    7 => 11176,
    8 => 11368,
    9 => 11709,
    10 => 11910,
    11 => 11943,
    12 => 12140,
    13 => 12196,
    14 => 18928,
    15 => 12494,
    17 => 13264,
    18 => 18742,
    19 => 18860,
  },
  "startingIndex": 0,
}
`;

exports[`readRefTable Expects to merge correctly the refTable of resources 2`] = `
Object {
  "maxIndex": 21,
  "offsets": Map {
    1 => 25091,
    2 => 19,
    3 => 224,
    4 => 12330,
    5 => 244,
    6 => 11154,
    7 => 11176,
    8 => 11368,
    9 => 11709,
    10 => 11910,
    11 => 11943,
    12 => 12140,
    13 => 12196,
    14 => 18948,
    15 => 12494,
    17 => 13264,
    18 => 18742,
    19 => 25016,
    20 => 19431,
    21 => 24878,
  },
  "startingIndex": 0,
}
`;

exports[`readRefTable Expects to merge correctly the refTable of resources 3`] = `
Object {
  "maxIndex": 24,
  "offsets": Map {
    1 => 15,
    2 => 4075,
    3 => 265,
    4 => 302,
    5 => 27161,
    6 => 55428,
    7 => 72070,
    8 => 378,
    9 => 592,
    10 => 764,
    11 => 4338,
    12 => 4394,
    13 => 4443,
    14 => 26241,
    15 => 26484,
    16 => 26817,
    17 => 27302,
    18 => 54210,
    19 => 54444,
    20 => 55060,
    21 => 55560,
    22 => 71288,
    23 => 71522,
    24 => 71742,
  },
  "startingIndex": 0,
}
`;

exports[`readRefTable Expects to merge correctly the refTable of resources 4`] = `
Object {
  "maxIndex": 62,
  "offsets": Map {
    1 => 53230,
    2 => 19,
    3 => 1785,
    4 => 1806,
    5 => 53757,
    6 => 53068,
    7 => 54181,
    8 => 54601,
    9 => 55043,
    10 => 55415,
    11 => 55817,
    12 => 57359,
    13 => 57859,
    14 => 58430,
    15 => 52999,
    16 => 58970,
    17 => 59969,
    18 => 60584,
    19 => 61203,
    20 => 61822,
    21 => 62441,
    22 => 62885,
    23 => 63408,
    24 => 53500,
    25 => 16564,
    26 => 28824,
    27 => 28847,
    28 => 29043,
    29 => 29414,
    30 => 29642,
    31 => 50764,
    32 => 50787,
    33 => 50978,
    34 => 51493,
    35 => 51848,
    36 => 52014,
    37 => 53132,
    38 => 53600,
    39 => 54024,
    40 => 54445,
    41 => 54886,
    42 => 55259,
    43 => 55631,
    44 => 57203,
    45 => 57702,
    46 => 58071,
    47 => 58274,
    48 => 58781,
    49 => 59610,
    50 => 59813,
    51 => 60225,
    52 => 60428,
    53 => 60844,
    54 => 61047,
    55 => 61463,
    56 => 61666,
    57 => 62082,
    58 => 62285,
    59 => 62701,
    60 => 63251,
    61 => 63625,
    62 => 63939,
  },
  "startingIndex": 0,
}
`;

exports[`readRefTable Expects to merge correctly the refTable of resources 5`] = `
Object {
  "maxIndex": 18,
  "offsets": Map {
    1 => 16964,
    2 => 19,
    3 => 224,
    4 => 12330,
    5 => 244,
    6 => 11154,
    7 => 11176,
    8 => 11368,
    9 => 11709,
    10 => 11910,
    11 => 11943,
    12 => 12140,
    13 => 12196,
    14 => 16880,
    15 => 12494,
    16 => 13264,
    17 => 16694,
    18 => 16812,
  },
  "startingIndex": 0,
}
`;

exports[`readRefTable Expects to merge correctly the refTable of resources 6`] = `
Object {
  "maxIndex": 13,
  "offsets": Map {
    1 => 4163,
    2 => 4098,
    3 => 4077,
    4 => 3812,
    5 => 3745,
    6 => 3619,
    7 => 15,
    8 => 3504,
    9 => 3439,
    10 => 4001,
    11 => 3932,
    12 => 3957,
    13 => 3982,
  },
  "startingIndex": 0,
}
`;

exports[`readRefTable Expects to merge correctly the refTable of resources 7`] = `
Object {
  "maxIndex": 15,
  "offsets": Map {
    1 => 11997,
    2 => 19,
    3 => 224,
    4 => 12330,
    5 => 244,
    6 => 11154,
    7 => 11176,
    8 => 11368,
    9 => 11709,
    10 => 11910,
    11 => 11943,
    12 => 12140,
    13 => 12196,
    14 => 12429,
    15 => 12494,
  },
  "startingIndex": 0,
}
`;
